In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import json

In [2]:
iris_df = pd.read_csv("iris.csv")
iris_df.head()

sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [3]:
j = """{
    "session_name": "test",
    "session_description": "test",
    "design_state_data": {

      "session_info" : {
        "project_id": "1",
        "experiment_id": "kkkk-11",
        "dataset":"iris_modified.csv",
        "session_name": "test",
        "session_description": "test"
        },

      "target": {
        "prediction_type": "Regression",
        "target": "petal_width",
        "type":"regression",
        "partitioning": true
      },
      "train": {
        "policy": "Split the dataset",
        "time_variable": "sepal_length",
        "sampling_method": "No sampling(whole data)",
        "split": "Randomly",
        "k_fold": false,
        "train_ratio": 0,
        "random_seed": 0
      },
      "metrics": {
        "optomize_model_hyperparameters_for": "AUC",
        "optimize_threshold_for": "F1 Score",
        "compute_lift_at": 0,
        "cost_matrix_gain_for_true_prediction_true_result": 1,
        "cost_matrix_gain_for_true_prediction_false_result": 0,
        "cost_matrix_gain_for_false_prediction_true_result": 0,
        "cost_matrix_gain_for_false_prediction_false_result": 0
      },
      "feature_handling": {
        "sepal_length": {
          "feature_name": "sepal_length",
          "is_selected": true,
          "feature_variable_type": "numerical",
          "feature_details": {
            "numerical_handling": "Keep as regular numerical feature",
            "rescaling": "No rescaling",
            "make_derived_feats": false,
            "missing_values": "Impute",
            "impute_with": "Average of values",
            "impute_value": 0
          }
        },
        "sepal_width": {
          "feature_name": "sepal_width",
          "is_selected": true,
          "feature_variable_type": "numerical",
          "feature_details": {
            "numerical_handling": "Keep as regular numerical feature",
            "rescaling": "No rescaling",
            "make_derived_feats": false,
            "missing_values": "Impute",
            "impute_with": "custom",
            "impute_value": -1
          }
        },
        "petal_length": {
          "feature_name": "petal_length",
          "is_selected": true,
          "feature_variable_type": "numerical",
          "feature_details": {
            "numerical_handling": "Keep as regular numerical feature",
            "rescaling": "No rescaling",
            "make_derived_feats": false,
            "missing_values": "Impute",
            "impute_with": "Average of values",
            "impute_value": 0
          }
        },
        "petal_width": {
          "feature_name": "petal_width",
          "is_selected": true,
          "feature_variable_type": "numerical",
          "feature_details": {
            "numerical_handling": "Keep as regular numerical feature",
            "rescaling": "No rescaling",
            "make_derived_feats": false,
            "missing_values": "Impute",
            "impute_with": "custom",
            "impute_value": -2
          }
        },
        "species": {
          "feature_name": "species",
          "is_selected": true,
          "feature_variable_type": "text",
          "feature_details": {
            "text_handling": "Tokenize and hash",
            "hash_columns": 0
          }
        }
      },
      "feature_generation": {
        "linear_interactions": [["petal_length", "sepal_width"]],
        "linear_scalar_type": "robust",
        "polynomial_interactions": [
          "petal_length/sepal_width",
          "petal_width/species"
        ],
        "explicit_pairwise_interactions": [
          "sepal_width/sepal_length",
          "petal_width/sepal_length"
        ]
      },
      "feature_reduction": {
        "feature_reduction_method": "Tree-based",
        "num_of_features_to_keep": "4",
        "num_of_trees": "5",
        "depth_of_trees": "6"
      },
      "hyperparameters": {
        "stratergy": "Grid Search",
        "shuffle_grid": true,
        "random_state": 1,
        "max_iterations": 2,
        "max_search_time": 3,
        "parallelism": 5,
        "cross_validation_stratergy": "Time-based K-fold(with overlap)",
        "num_of_folds": 6,
        "split_ratio": 0,
        "stratified": true
      },
      "weighting_stratergy": {
        "weighting_stratergy_method": "Sample weights",
        "weighting_stratergy_weight_variable": "petal_length"
      },
      "probability_calibration": {
        "probability_calibration_method": "Sigmoid - Platt Scaling"
      },
      "algorithms": {
        "RandomForestClassifier": {
          "model_name": "Random Forest Classifier",
          "is_selected": false,
          "min_trees": 10,
          "max_trees": 30,
          "feature_sampling_statergy": "Default",
          "min_depth": 20,
          "max_depth": 30,
          "min_samples_per_leaf_min_value": 5,
          "min_samples_per_leaf_max_value": 50,
          "parallelism": 0
        },
        "RandomForestRegressor": {
          "model_name": "Random Forest Regressor",
          "is_selected": true,
          "min_trees": 10,
          "max_trees": 20,
          "feature_sampling_statergy": "Default",
          "min_depth": 20,
          "max_depth": 25,
          "min_samples_per_leaf_min_value": 5,
          "min_samples_per_leaf_max_value": 10,
          "parallelism": 0
        },
        "GBTClassifier": {
          "model_name": "Gradient Boosted Trees",
          "is_selected": false,
          "num_of_BoostingStages": [67, 89],
          "feature_sampling_statergy": "Fixed number",
          "learningRate": [],
          "use_deviance": true,
          "use_exponential": false,
          "fixed_number": 22,
          "min_subsample": 1,
          "max_subsample": 2,
          "min_stepsize":0.1,
          "max_stepsize":0.5, 
          "min_iter":20,
          "max_iter":40,
          "min_depth":5,
          "max_depth":7

        },
        "GBTRegressor": {
          "model_name": "Gradient Boosted Trees",
          "is_selected": false,
          "num_of_BoostingStages": [67, 89],
          "feature_sampling_statergy": "Fixed number",
          "use_deviance": true,
          "use_exponential": false,
          "fixed_number": 22,
          "min_subsample": 1,
          "max_subsample": 2,
          "min_stepsize":0.1,
          "max_stepsize":0.5, 
          "min_iter":20,
          "max_iter":40,
          "min_depth":5,
          "max_depth":7
        },
        "LinearRegression": {
          "model_name": "LinearRegression",
          "is_selected": false,
          "parallelism": 2,
          "min_iter":30,
          "max_iter":50,
          "min_regparam":0.5,
          "max_regparam":0.8,
          "min_elasticnet":0.5,
          "max_elasticnet":0.8
        },
        "LogisticRegression": {
          "model_name": "LogisticRegression",
          "is_selected": false,
          "parallelism": 2,
          "min_iter":30,
          "max_iter":50,
          "min_regparam":0.5,
          "max_regparam":0.8,
          "min_elasticnet":0.5,
          "max_elasticnet":0.8
        },
        "RidgeRegression": {
          "model_name": "RidgeRegression",
          "is_selected": false,
          "regularization_term": "Specify values to test",
          "min_iter":30,
          "max_iter":50,
          "min_regparam":0.5,
          "max_regparam":0.8
        },
        "LassoRegression": {
          "model_name": "Lasso Regression",
          "is_selected": false,
          "regularization_term": "Specify values to test",
          "min_iter":30,
          "max_iter":50,
          "min_regparam":0.5,
          "max_regparam":0.8
        },
        "ElasticNetRegression": {
          "model_name": "Lasso Regression",
          "is_selected": false,
          "regularization_term": "Specify values to test",
          "min_iter":30,
          "max_iter":50,
          "min_regparam":0.5,
          "max_regparam":0.8,
          "min_elasticnet":0.5,
          "max_elasticnet":0.8
        },
        "xg_boost": {
          "model_name": "XG Boost",
          "is_selected": false,
          "use_gradient_boosted_tree": true,
          "dart": true,
          "tree_method": "",
          "random_state": 0,
          "max_num_of_trees": 0,
          "early_stopping": true,
          "early_stopping_rounds": 2,
          "max_depth_of_tree": [56, 89], 
          "learningRate": [89, 76],
          "l1_regularization": [77],
          "l2_regularization": [78],
          "gamma": [68],
          "min_child_weight": [67],
          "sub_sample": [67],
          "col_sample_by_tree": [67],
          "replace_missing_values": false,
          "parallelism": 0
        },
        "DecisionTreeRegressor": {
          "model_name": "Decision Tree",
          "is_selected": false,
          "min_depth":4,
          "max_depth": 7,
          "use_gini": false,
          "use_entropy": true,
          "min_samples_per_leaf": [12, 6],
          "use_best": true,
          "use_random": true
        },
        "DecisionTreeClassifier": {
          "model_name": "Decision Tree",
          "is_selected": false,
          "min_depth":4,
          "max_depth": 7,
          "use_gini": false,
          "use_entropy": true,
          "min_samples_per_leaf": [12, 6],
          "use_best": true,
          "use_random": true
        },
        "SVM": {
          "model_name": "Support Vector Machine",
          "is_selected": false,
          "linear_kernel": true,
          "rep_kernel": true,
          "polynomial_kernel": true,
          "sigmoid_kernel": true,
          "c_value": [566, 79],
          "auto": true,
          "scale": true,
          "custom_gamma_values": true,
          "tolerance": 7,
          "max_iterations": 7
        },
        "SGD": {
          "model_name": "Stochastic Gradient Descent",
          "is_selected": false,
          "use_logistics": true,
          "use_modified_hubber_loss": false,
          "max_iterations": false,
          "tolerance": 56,
          "use_l1_regularization": "on",
          "use_l2_regularization": "on",
          "use_elastic_net_regularization": true,
          "alpha_value": [79, 56],
          "parallelism": 1
        },
        "KNN": {
          "model_name": "KNN",
          "is_selected": false,
          "k_value": [78],
          "distance_weighting": true,
          "neighbour_finding_algorithm": "Automatic",
          "random_state": 0,
          "p_value": 0
        },
        "extra_random_trees": {
          "model_name": "Extra Random Trees",
          "is_selected": false,
          "num_of_trees": [45, 489],
          "feature_sampling_statergy": "Square root and Logarithm",
          "max_depth": [12, 45],
          "min_samples_per_leaf": [78, 56],
          "parallelism": 3
        },
        "neural_network": {
          "model_name": "Neural Network",
          "is_selected": false,
          "hidden_layer_sizes": [67, 89],
          "activation": "",
          "alpha_value": 0,
          "max_iterations": 0,
          "convergence_tolerance": 0,
          "early_stopping": true,
          "solver": "ADAM",
          "shuffle_data": true,
          "initial_learning_rate": 0,
          "automatic_batching": true,
          "beta_1": 0,
          "beta_2": 0,
          "epsilon": 0,
          "power_t": 0,
          "momentum": 0,
          "use_nesterov_momentum": false
        }
      }
    }
  }"""

In [4]:
data = json.loads(j)

### Target

In [5]:
target = data["design_state_data"]["target"]

In [6]:
target

{'prediction_type': 'Regression',
 'target': 'petal_width',
 'type': 'regression',
 'partitioning': True}

### Target Variable

In [7]:
target_variable = target["target"]

In [8]:
target_variable

'petal_width'

### Prediction Type

In [9]:
prediction_type = target["prediction_type"]
prediction_type

'Regression'

### Type of Regression

In [10]:
type_of_reg = target["type"]
type_of_reg

'regression'

### Feature Handling

In [11]:
feature_handling = data["design_state_data"]["feature_handling"]

In [12]:
feature_handling

{'sepal_length': {'feature_name': 'sepal_length',
  'is_selected': True,
  'feature_variable_type': 'numerical',
  'feature_details': {'numerical_handling': 'Keep as regular numerical feature',
   'rescaling': 'No rescaling',
   'make_derived_feats': False,
   'missing_values': 'Impute',
   'impute_with': 'Average of values',
   'impute_value': 0}},
 'sepal_width': {'feature_name': 'sepal_width',
  'is_selected': True,
  'feature_variable_type': 'numerical',
  'feature_details': {'numerical_handling': 'Keep as regular numerical feature',
   'rescaling': 'No rescaling',
   'make_derived_feats': False,
   'missing_values': 'Impute',
   'impute_with': 'custom',
   'impute_value': -1}},
 'petal_length': {'feature_name': 'petal_length',
  'is_selected': True,
  'feature_variable_type': 'numerical',
  'feature_details': {'numerical_handling': 'Keep as regular numerical feature',
   'rescaling': 'No rescaling',
   'make_derived_feats': False,
   'missing_values': 'Impute',
   'impute_with': '

### Included Features

In [13]:
features = []
print("Included features are  = ",end = " ")
for feature in iris_df.columns:
    if feature_handling[f"{feature}"]["is_selected"] == True:
        features.append(feature)
print(features)

Included features are  =  ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']


### Imputing rules for each feature

In [14]:
print("Imputation for each feature(incase of missing values):")
for feature in features:
    if feature == 'species':
        continue
    print(f"{feature}", end = ": ")
    print("{}".format(feature_handling[f"{feature}"]["feature_details"]["missing_values"]))

Imputation for each feature(incase of missing values):
sepal_length: Impute
sepal_width: Impute
petal_length: Impute
petal_width: Impute


#### Sepal Length

In [15]:
feature_handling["sepal_length"]

{'feature_name': 'sepal_length',
 'is_selected': True,
 'feature_variable_type': 'numerical',
 'feature_details': {'numerical_handling': 'Keep as regular numerical feature',
  'rescaling': 'No rescaling',
  'make_derived_feats': False,
  'missing_values': 'Impute',
  'impute_with': 'Average of values',
  'impute_value': 0}}

In [16]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [17]:
print("number of null values in feature 'sepal_length': ",iris_df["sepal_length"].isnull().sum())

number of null values in feature 'sepal_length':  0


Hence no imputation required on sepal_length

### Sepal_width

In [18]:
feature_handling["sepal_width"]

{'feature_name': 'sepal_width',
 'is_selected': True,
 'feature_variable_type': 'numerical',
 'feature_details': {'numerical_handling': 'Keep as regular numerical feature',
  'rescaling': 'No rescaling',
  'make_derived_feats': False,
  'missing_values': 'Impute',
  'impute_with': 'custom',
  'impute_value': -1}}

In [19]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [20]:
iris_df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [21]:
print("number of null values in feature 'sepal_width': ",iris_df["sepal_width"].isnull().sum())

number of null values in feature 'sepal_width':  0


Hence no imputation required on sepal_width

### Petal Length

In [22]:
feature_handling["petal_length"]

{'feature_name': 'petal_length',
 'is_selected': True,
 'feature_variable_type': 'numerical',
 'feature_details': {'numerical_handling': 'Keep as regular numerical feature',
  'rescaling': 'No rescaling',
  'make_derived_feats': False,
  'missing_values': 'Impute',
  'impute_with': 'Average of values',
  'impute_value': 0}}

In [23]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [24]:
iris_df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [25]:
print("number of null values in feature 'petal_length': ",iris_df["petal_length"].isnull().sum())

number of null values in feature 'petal_length':  0


Hence no imputation is required for feature 'petal_length'

### Petal_width

In [26]:
feature_handling["petal_width"]

{'feature_name': 'petal_width',
 'is_selected': True,
 'feature_variable_type': 'numerical',
 'feature_details': {'numerical_handling': 'Keep as regular numerical feature',
  'rescaling': 'No rescaling',
  'make_derived_feats': False,
  'missing_values': 'Impute',
  'impute_with': 'custom',
  'impute_value': -2}}

In [27]:
iris_df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [28]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [29]:
print("number of null values in feature 'petal_width': ",iris_df["petal_width"].isnull().sum())

number of null values in feature 'petal_width':  0


Hence no imputation is required in petal_width

### Species

In [30]:
feature_handling["species"]

{'feature_name': 'species',
 'is_selected': True,
 'feature_variable_type': 'text',
 'feature_details': {'text_handling': 'Tokenize and hash', 'hash_columns': 0}}

In [31]:
iris_df.species.unique()

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

### Encoding for Species

In [32]:
iris_df1 = pd.get_dummies(iris_df)

In [33]:
iris_df1.head()

sepal_length  sepal_width  petal_length  petal_width  species_Iris-setosa  \
0           5.1          3.5           1.4          0.2                    1   
1           4.9          3.0           1.4          0.2                    1   
2           4.7          3.2           1.3          0.2                    1   
3           4.6          3.1           1.5          0.2                    1   
4           5.0          3.6           1.4          0.2                    1   

   species_Iris-versicolor  species_Iris-virginica  
0                        0                       0  
1                        0                       0  
2                        0                       0  
3                        0                       0  
4                        0                       0

### Model choosing

In [34]:
prediction_type

'Regression'

In [35]:
type_of_reg

'regression'

In [37]:
models = []

for k in data['design_state_data']['algorithms'].keys():
    # print(k)
    if data['design_state_data']['algorithms'][f"{k}"]['is_selected'] == True:
        models.append(k)

In [38]:
models

['RandomForestRegressor']

Hence, we will use RandomForestRegressor

Now, get the details about the hyperparameters of the model

In [46]:
data['design_state_data']['algorithms'][f"{models[0]}"]

{'model_name': 'Random Forest Regressor',
 'is_selected': True,
 'min_trees': 10,
 'max_trees': 20,
 'feature_sampling_statergy': 'Default',
 'min_depth': 20,
 'max_depth': 25,
 'min_samples_per_leaf_min_value': 5,
 'min_samples_per_leaf_max_value': 10,
 'parallelism': 0}

### Model Building

In [53]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators = 10, max_depth = 20, min_samples_leaf = 5)

### Data processing

In [48]:
x = iris_df.drop(f"{target_variable}", axis = 1)
y = iris_df[f"{target_variable}"]

In [49]:
x

sepal_length  sepal_width  petal_length         species
0             5.1          3.5           1.4     Iris-setosa
1             4.9          3.0           1.4     Iris-setosa
2             4.7          3.2           1.3     Iris-setosa
3             4.6          3.1           1.5     Iris-setosa
4             5.0          3.6           1.4     Iris-setosa
..            ...          ...           ...             ...
145           6.7          3.0           5.2  Iris-virginica
146           6.3          2.5           5.0  Iris-virginica
147           6.5          3.0           5.2  Iris-virginica
148           6.2          3.4           5.4  Iris-virginica
149           5.9          3.0           5.1  Iris-virginica

[150 rows x 4 columns]

In [50]:
y

0      0.2
1      0.2
2      0.2
3      0.2
4      0.2
      ... 
145    2.3
146    1.9
147    2.0
148    2.3
149    1.8
Name: petal_width, Length: 150, dtype: float64

### Handling Categorical variable

In [51]:
x2 = pd.get_dummies(x)
x2.head()

sepal_length  sepal_width  petal_length  species_Iris-setosa  \
0           5.1          3.5           1.4                    1   
1           4.9          3.0           1.4                    1   
2           4.7          3.2           1.3                    1   
3           4.6          3.1           1.5                    1   
4           5.0          3.6           1.4                    1   

   species_Iris-versicolor  species_Iris-virginica  
0                        0                       0  
1                        0                       0  
2                        0                       0  
3                        0                       0  
4                        0                       0

Drop species_Iris_virginica column to avoid dummy variable trap

In [52]:
x2 = x2.drop('species_Iris-virginica', axis = 1)
x2

sepal_length  sepal_width  petal_length  species_Iris-setosa  \
0             5.1          3.5           1.4                    1   
1             4.9          3.0           1.4                    1   
2             4.7          3.2           1.3                    1   
3             4.6          3.1           1.5                    1   
4             5.0          3.6           1.4                    1   
..            ...          ...           ...                  ...   
145           6.7          3.0           5.2                    0   
146           6.3          2.5           5.0                    0   
147           6.5          3.0           5.2                    0   
148           6.2          3.4           5.4                    0   
149           5.9          3.0           5.1                    0   

     species_Iris-versicolor  
0                          0  
1                          0  
2                          0  
3                          0  
4                          0  
..                       ...  
145                        0  
146                        0  
147                        0  
148                        0  
149                        0  

[150 rows x 5 columns]

### Train Test Split

In [54]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x2, y, test_size = 0.2, random_state = 10)

### Train - cv split

In [58]:
x_train, x_cv, y_train, y_cv = train_test_split(x_train, y_train, test_size = 0.2, random_state = 10)

In [59]:
x_train.head()

sepal_length  sepal_width  petal_length  species_Iris-setosa  \
56           6.3          3.3           4.7                    0   
83           6.0          2.7           5.1                    0   
82           5.8          2.7           3.9                    0   
17           5.1          3.5           1.4                    1   
88           5.6          3.0           4.1                    0   

    species_Iris-versicolor  
56                        1  
83                        1  
82                        1  
17                        0  
88                        1

In [60]:
y_train.head()

56    1.6
83    1.6
82    1.2
17    0.3
88    1.3
Name: petal_width, dtype: float64

### Training

In [61]:
model.fit(x_train, y_train)

RandomForestRegressor(max_depth=20, min_samples_leaf=5, n_estimators=10)

### Testing against cross-validation set

In [62]:
model.score(x_cv, y_cv)

0.92915238813152

Hence we get an accuracy of 92.915% on cv set

### Tuning The hyper parameters

In [64]:
from sklearn.model_selection import GridSearchCV

In [65]:
algo = {
    'n_estimators':range(10,21),
    'max_depth': range(20,26),
    'min_samples_leaf': range(5,11)
}

In [66]:
clf = GridSearchCV(RandomForestRegressor, algo)

In [67]:
clf.fit(x_train, y_train)

GridSearchCV(estimator=RandomForestRegressor(max_depth=20, min_samples_leaf=5,
                                             n_estimators=10),
             param_grid={'max_depth': range(20, 26),
                         'min_samples_leaf': range(5, 11),
                         'n_estimators': range(10, 21)})

In [69]:
clf.cv_results_

{'mean_fit_time': array([0.02831774, 0.02517948, 0.02740693, 0.02920737, 0.03037262,
        0.03286128, 0.03442893, 0.036482  , 0.03803596, 0.04015317,
        0.04179244, 0.0228929 , 0.02547216, 0.0272975 , 0.02837462,
        0.03051472, 0.03424315, 0.03230157, 0.03369594, 0.03666449,
        0.03977895, 0.04156713, 0.02299519, 0.02491064, 0.02765808,
        0.02904949, 0.03005795, 0.0326335 , 0.03361368, 0.03554158,
        0.03557825, 0.03991938, 0.04167061, 0.02370114, 0.02315788,
        0.02544956, 0.02641749, 0.0274251 , 0.0288219 , 0.033638  ,
        0.03282962, 0.03968129, 0.04171329, 0.04110751, 0.02310658,
        0.02510414, 0.0268065 , 0.02931795, 0.0338593 , 0.03356466,
        0.03520193, 0.03694348, 0.03971581, 0.04025311, 0.04256148,
        0.02344141, 0.02576618, 0.02676935, 0.02845321, 0.03157239,
        0.03312759, 0.03506789, 0.03616757, 0.04097648, 0.04078093,
        0.04104118, 0.02394385, 0.02385378, 0.02698979, 0.02899857,
        0.03152437, 0.03397713,

In [73]:
print('best score = ', clf.best_score_)
print('best params = ', clf.best_params_)

best score =  0.9375786569804883
best params =  {'max_depth': 20, 'min_samples_leaf': 5, 'n_estimators': 20}


### Training model using best params

In [75]:
model2 = RandomForestRegressor(max_depth = 20, min_samples_leaf = 5, n_estimators = 20)

In [76]:
model2.fit(x_train, y_train)

RandomForestRegressor(max_depth=20, min_samples_leaf=5, n_estimators=20)

In [86]:
print("Score on Training Set = ", model2.score(x_train, y_train)*100, "%")
print("Score on CV Set = ", model2.score(x_cv, y_cv)*100, "%")

Score on Training Set =  96.1119430814793 %
Score on CV Set =  93.03380421097496 %


## Test Score

In [87]:
print("Score on Test Set = ", model2.score(x_test, y_test)*100, "%")

Score on Test Set =  97.56983176767012 %


In [80]:
import pickle

In [85]:
with open("model", 'wb') as f:
    pickle.dump(model2, f)

### Conclusion

The RandomForestRegressor with the best Parameters gives(Approximately) a Train score of 96.112%, CV score of 93.034% and Test Score of 97.57%